### PR TITLE
Update training docs and code

### LINKED ISSUE #68: Cannot access conllparser.py
Hi
I am not able to access the conllparser.py script here: https://github.com/huggingface/neuralcoref/blob/master/neuralcoref/conllparser.py

I find no folder named blob in the repository.

Could you please check?

Thanks,
Basil.

### LINKED ISSUE #89: prepare the data error
"python -m neuralcoref.conllparser --path ./data/train/"  returns an error message "AttributeError: 'ConllDoc' object has no attribute '_extract_mentions'"

The traceback is as below, 
🌋 Parsing utterances and filling docs
0it [00:00, ?it/s]
Traceback (most recent call last):
  File ".../python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File ".../python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File ".../neuralcoref/neuralcoref/conllparser.py", line 711, in <module>
    CORPUS.read_corpus(args.path)
  File ".../neuralcoref/neuralcoref/conllparser.py", line 618, in read_corpus
    use_gold_mentions=self.use_gold_mentions)
  File ".../neuralcoref/neuralcoref/conllparser.py", line 318, in add_conll_utterance
    self._extract_mentions(parsed, len(self.utterances), self.n_sents, self.speakers[speaker_id])

Method _extract_mentions is missing from class ConllDoc or class Document

### LINKED ISSUE #119: The training instructions no longer correspond to the actual package structure
I'm trying to train a model (with Conll and Ontonotes files) using the instructions from  : https://github.com/huggingface/neuralcoref/blob/master/neuralcoref/train/training.md

Unfortunately the current code doesn't reflect those instructions. In the first place the conllparser.py module is now under "train" directory so : python -m neuralcoref.conllparser ...command fails.
Secondly, after I move all the files back in "neuralcoref" and execute the training command this fails with the error : 
self._extract_mentions(parsed, len(self.utterances), self.n_sents, self.speakers[speaker_id])
AttributeError: 'ConllDoc' object has no attribute '_extract_mentions'
Indeed the class ConllDoc has no such attribute.
 Is it still possible to create a model? And if yes how?

Thanks!
Eduard

### LINKED ISSUE #131: 'ConllDoc' object has no attribute '_extract_mentions' error
I already had problem with neuralcoref.conllparser not found module, however, I overcame this challenge by fix directly the python file and run directly the python file.

When I ran the conllparser.py file, I continued to have this error. I have checked and see that there is no definition of the extract_mention method in ConllDoc class. The code seems to incomplete I guess.

Parsing utterances and filling docs
  0%|                                                | 0/238192 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "conllparser.py", line 711, in <module>
    CORPUS.read_corpus(args.path)
  File "conllparser.py", line 618, in read_corpus
    use_gold_mentions=self.use_gold_mentions)
  File "conllparser.py", line 318, in add_conll_utterance
    self._extract_mentions(parsed, len(self.utterances), self.n_sents, self.speakers[speaker_id])
AttributeError: 'ConllDoc' object has no attribute '_extract_mentions'

### LINKED ISSUE #187: Where is the conll parser?
In the [instructions](https://github.com/huggingface/neuralcoref/blob/master/neuralcoref/train/training.md) there is a reference to a conll parser and conll processing scripts, but those links are dead. They have been removed but it's not clear to me why.

### LINKED ISSUE #196: AttributeError: 'ConllDoc' object has no attribute '_extract_mentions'
Excuse me, has anyone solved the following problem?
AttributeError: 'ConllDoc' object has no attribute '_extract_mentions'

Cann‘t generate the data of the training model. Does anyone have .npy data under the numpy folder? I want to convert the language to Chinese.

### PR DESCRIPTION
**Pull request by svlandeg**
A recent refactor of the code seems to have left a few inconsistencies in the code & documentation. Trying to fix those here as an intermediate fix to enable retraining of the models with the "old" code. A further refactor to improve speed, as Thomas was planning, is probably next on the TODO-list.
 
## Fixes 
* Updates code to work with PyTorch 1.3.0
* Fixing paths after moving a lot of code from root `neuralcoref` to subdir `training`
* Fixed reference to `self._extract_mentions` in `conllparser.py` (@thomwolf : could you review this bit?)
    * Fixes #89
    * Fixes #119
    * Fixes #131
    * Fixes #196 
 
## New functionality 
* Try not just `en_core_web_sm`, but any model in the list `['en_core_web_sm', 'en', 'en_core_web_md', 'en_core_web_lg']`, giving preference to the largest model
* Setting `blacklist` and `use_gold_mentions` as script parameters

## Documentation
* `training.md` : step-by-step details on processing CONLL & OntoNotes data
    * Fixes #68
    * Fixes #187

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,6 +15,7 @@ neuralcoref/*.html
 Profile.prof
 .vscode
 .sass-cache
+.idea/*
```
Comment (thomwolf) [inline .gitignore]: 👍

File: neuralcoref/train/conll_processing_script/compile_coref_data.sh [hunk 2]
```diff
@@ -12,7 +12,7 @@
 ONTONOTES_PATH=$1
 
 if [ ! -n "$ONTONOTES_PATH" ] ; then
-  echo "USAGE: ./scripts/compile_coref_data.sh /path/to/ontonotes/data"
+  echo "USAGE: ./compile_coref_data.sh /path/to/ontonotes/data"
```
Comment (thomwolf) [inline neuralcoref/train/conll_processing_script/compile_coref_data.sh]: In the future use something like this: https://github.com/ns-moosavi/coval

File: neuralcoref/train/conllparser.py [hunk 3]
```diff
@@ -315,7 +313,9 @@ def add_conll_utterance(self, parsed, tokens, corefs, speaker_id, use_gold_menti
                 self.mentions.append(mention)
                 #            print("mention: ", mention, "label", mention.gold_label)
         else:
-            self._extract_mentions(parsed, len(self.utterances), self.n_sents, self.speakers[speaker_id])
+            mentions_spans = extract_mentions_spans(doc=parsed, blacklist=self.blacklist)
```
Comment (thomwolf) [inline neuralcoref/train/conllparser.py]: Check if you want to use it (what's in blacklist)

Comment (thomwolf) [reply neuralcoref/train/conllparser.py]: Otherwise looks good to me

File: neuralcoref/train/conllparser.py [hunk 4]
```diff
@@ -594,13 +594,20 @@ def read_corpus(self, data_path, debug=False):
...
                                       conll=CONLL_GENRES[name[:2]]))
         print("🌋 Loading spacy model")
-        try:
-            spacy.info('en_core_web_sm')
-            model = 'en_core_web_sm'
-        except IOError:
-            print("No spacy 2 model detected, using spacy1 'en' model")
-            spacy.info('en')
-            model = 'en'
+        model_options = ['en_core_web_sm', 'en', 'en_core_web_md', 'en_core_web_lg']
```
Comment (thomwolf) [inline neuralcoref/train/conllparser.py]: Favor the best model for parsing!

File: neuralcoref/train/conllparser.py [hunk 5]
```diff
@@ -618,10 +625,10 @@ def read_corpus(self, data_path, debug=False):
                                                   use_gold_mentions=self.use_gold_mentions)
 
     def build_and_gather_multiple_arrays(self, save_path):
-        print("🌋 Extracting mentions features")
+        print("🌋 Extracting mentions features with {} job(s)".format(self.n_jobs))
         parallel_process(self.docs, set_feats, n_jobs=self.n_jobs)
```
Comment (thomwolf) [inline neuralcoref/train/conllparser.py]: Switch to the cython version one day

File: neuralcoref/train/conllparser.py [hunk 6]
```diff
@@ -652,20 +659,26 @@ def build_and_gather_multiple_arrays(self, save_path):
...
-                array = np.array(gathering_dict[feature])
+                array = np.array(feature_data)
                 if array.ndim == 1:
                     array = np.expand_dims(array, axis=1)
             else:
-                array = np.stack(gathering_dict[feature])
+                array = np.stack(feature_data)
             # check_numpy_array(feature, array, n_mentions_list)
             print("Saving numpy", feature, "size", array.shape)
             np.save(save_path + feature, array)
```
Comment (thomwolf) [inline neuralcoref/train/conllparser.py]: Maybe if the files are too big, you can change the numpy dtype (from long to int)

File: neuralcoref/train/evaluator.py [hunk 7]
```diff
@@ -134,7 +133,8 @@ def display_clusters(self, doc_idx=None):
     ########################
     def get_max_score(self, batch, debug=False):
         inputs, mask = batch
-        inputs = tuple(Variable(i, volatile=True) for i in inputs)
+        with torch.no_grad():
+            inputs = tuple(i for i in inputs)
```
Comment (thomwolf) [inline neuralcoref/train/evaluator.py]: Two stuff here:
- Can remove this line `inputs = tuple(i for i in inputs)`
- the ` with torch.no_grad():` context will go around the forward pass and it should be like this:
```python
with torch.no_grad():
    scores = self.model(inputs, concat_axis=1)
    scores.masked_fill_(mask, -float('Inf'))
    _, max_idx = scores.max(dim=1) # We may want to weight the single score with coref.greedyness
```

File: neuralcoref/train/evaluator.py [hunk 8]
```diff
@@ -134,7 +133,8 @@ def display_clusters(self, doc_idx=None):
...
         if self.cuda:
             inputs = tuple(i.cuda() for i in inputs)
```
Comment (thomwolf) [inline neuralcoref/train/evaluator.py]: if you do distributed training you will store the current device in a variable and use `tensor.to(device)`
Example of distributed training: https://github.com/huggingface/transformers/blob/master/examples/run_glue.py#L426
How to run a distributed training setup: https://github.com/huggingface/transformers/tree/master/examples#distributed-training
More information on training that are too long and how to solve that: https://medium.com/huggingface/training-larger-batches-practical-tips-on-1-gpu-multi-gpu-distributed-setups-ec88c3e51255

**Comment by thomwolf** [comment]
Possible next step:
- switch the evaluation scripts to the python ones => speed up the evaluation a lot
- switch the mention extraction and feature computation to the cython neural coref => speed up data processing a lot

**Comment by thomwolf** [comment]
- add a clean script to convert pytorch weights in thinc (probably using numpy as a conversion format)

File: neuralcoref/train/conllparser.py (hunk 3 above)
Comment (svlandeg) [reply neuralcoref/train/conllparser.py]: it's not used by default, I made it an argparse parameter to be able to set it easily

**Comment by svlandeg** [comment]
Thanks a lot for the review and the tips Thomas! I've added all your pointers to the TODO list :-) This PR is virtually ready - will run the latest code once more to see if it trains correctly, then merge.

**Comment by thomwolf** [comment]
Awesome! Great work Sofie!

**Pull request merged**